### PR TITLE
Faster and more correct Symbol::findMemberFuzzyMatchConstant

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Speed up fuzzy constant match and fix a significant bug.
    
Fuzzy constant match called `Symbol::membersStableOrderSlow` to deterministically loop through a symbol's members. As the name implies, this method is expensive; it sorts members by their string names. When Sorbet generates many errors, this function call dominates runtime (by >80%).
    
However, this function call is not needed to ensure deterministic fuzzy matches! We can cheaply sort and tiebreak after finding the best matches in a given scope.
    
Unfortunately, I noticed that this method has some peculiar behavior:
* Rather than pick one best candidate per scope (as a comment at the top of the method says), it picks `n` semi-arbitrary best candidates within the scope that happen to beat "betterThan". This behavior creates a dependency between iteration order and the contents of `result`, and ensures that less-than-ideal matches are presented to the user.
* The function sorts the candidates to place the suggestions with the smallest edit distance first, __and then reverses the vector__ to place the largest edit distances first. As a result, Sorbet would recommend the _worst_ replacements first!
  * Look at `constant-fuzzy.out` to see how we used to recommend `Date` over `Integer` for `Intege` (!!!)
    
I have changed the behavior to:
* Pick the single best candidates in the current scope that beats out candidates from inner scopes, and sort them by symbol ID to tiebreak.
* Place the items with the smallest edit distance _first_ in the result vector without extraneous sorting.
    
From the changes in test output, I believe this will significantly improve Sorbet's autocorrect suggestions _and_ is a massive perf win when Sorbet generates many errors.

### Motivation

See https://github.com/sorbet/sorbet/pull/4239#issuecomment-853243842

### Test plan

See included automated tests.

### CODE REVIEW

File: core/Symbols.cc [hunk 1]
```diff
@@ -700,29 +701,40 @@ vector<Symbol::FuzzySearchResult> Symbol::findMemberFuzzyMatchConstant(const Glo
...
+                            scopeBest.emplace_back(best);
                         }
                     }
                 }
+                if (!scopeBest.empty()) {
+                    // NOTE: Iteration order over members is nondeterministic, so we use SymbolId to deterministically
+                    // order the recommendations from this scope.
+                    // We order in decreasing symbol ID order because `result` is later reversed and we want earlier
+                    // ID'd symbols to be recommended first.
+                    fast_sort(scopeBest, [&](auto lhs, auto rhs) -> bool { return lhs.symbol._id > rhs.symbol._id; });
```
Comment (froydnj) [inline core/Symbols.cc:726]: I don't have the rules for `auto` deduction in lambda lists memorized, but I'm fairly sure that here it resolves to `Symbol::FuzzySearchResult` and we're therefore copy-constructing two items for every comparison.  Even though that copy construction is cheap, I don't think we want to do that.  Can we please use `const auto &` for `lhs` and `rhs` here?

(I realize this is just following the existing practice for the previously-existing sort in this function, but I think it was wrong there, too.)

File: core/Symbols.cc [hunk 2]
```diff
@@ -758,10 +770,19 @@ vector<Symbol::FuzzySearchResult> Symbol::findMemberFuzzyMatchConstant(const Glo
                 }
             }
         }
+        // globalBest is nondeterministically ordered since iteration over `members()` is non deterministic.
+        // Everything in globalBest has the same edit distance, so we just have to sort by symbol ID to get a
+        // deterministic order.
+        // We order in decreasing symbol ID order because `result` is later reversed and we want earlier
+        // ID'd symbols to be recommended first.
+        fast_sort(globalBest, [&](auto lhs, auto rhs) -> bool { return lhs.symbol._id > rhs.symbol._id; });
```
Comment (froydnj) [inline core/Symbols.cc:778]: Same ask for `const auto &` on `lhs`/`rhs` here.

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JbBiG8WZg46uQR
→ https://go/builds/bui_JbBibe9ICny0KB

File: test/testdata/lsp/missing_typed_sigil.A.rbedited [hunk 3]
```diff
@@ -12,7 +12,7 @@ class Hello
 end
 
 def main
-  puts Gem.new
+  puts Hello.new
   #    ^^^^ error: Unable to resolve constant `Helo`
-  #    ^^^^ apply-code-action: [A] Replace with `Gem`
+  #    ^^^^ apply-code-action: [A] Replace with `Hello`
```
Comment (jvilk-stripe) [inline test/testdata/lsp/missing_typed_sigil.A.rbedited:17]: :chefskiss:

File: core/Symbols.cc [hunk 4]
```diff
@@ -700,29 +701,40 @@ vector<Symbol::FuzzySearchResult> Symbol::findMemberFuzzyMatchConstant(const Glo
...
                 }
+                if (!scopeBest.empty()) {
+                    // NOTE: Iteration order over members is nondeterministic, so we use SymbolId to deterministically
+                    // order the recommendations from this scope.
+                    // We order in decreasing symbol ID order because `result` is later reversed and we want earlier
+                    // ID'd symbols to be recommended first.
+                    fast_sort(scopeBest, [&](auto lhs, auto rhs) -> bool { return lhs.symbol._id > rhs.symbol._id; });
+                    for (auto item : scopeBest) {
+                        result.emplace_back(item);
+                    }
```
Comment (froydnj) [inline core/Symbols.cc:729]: I don't know whether the compiler would generate the same code here as with `auto &item`, but I think the latter is less confusing to reason about.  (Note that `auto &` is used when iterating over `globalBest`, below, so it doesn't seem completely unreasonable to use `auto &` here.)  Would you please use `auto &item` instead?

**Review by froydnj — APPROVED** [review]
Thank you for the comments in `findMemberFuzzyMatchConstant`!  They made things a lot easier to follow.

**Comment by jvilk-stripe** [comment]
I tested this change on Stripe's codebase. While tens of thousands of errors still slows down typechecking, it does complete in a finite amount of time and interrupting it to fix the issue seems to be reasonably responsive. The errors print out to the CLI significantly faster too.

**Pull request merged**